File: javascript/src/dispatch.js

const poly = {
    "new" : function (initvalues = null) {
        let methodTable = new Map()
        if (initvalues != null) {
            console.log("")
        }
        let fn = function (obj, ...args){
            let objType = typeof(obj)
            if (objType == "object"){
                let meta = Object.getPrototypeOf(obj)
                while (meta != null){
                    method = methodTable.get(meta)
                    if (!(method === undefined)){
                        return method(obj, ...args)
                    }
                }
                objType = "any"
            } else {
                method = methodTable.get(objType)
                if (!(method === undefined)){
                    return method(obj, ...args)
                } else {
                    throw Error("No methods found")
                }
            }

        }
        fn.methodTable = methodTable
        fn.setmethod = function (type, method){
            this.methodTable.set(type, method)
        }
        return fn
    },
    "register" : function (fn, type, method){
        fn.methodTable.set(type, method)
    },
}


const _Array = Object.create(null)

let array = Object.create(_Array)
array.data = [3,5,7]

const UnitRange = Object.create(_Array)
let range = Object.create(UnitRange)
range.length = 5

let iterate = poly.new()
iterate.setmethod( _Array, function (a){return a.data})
iterate.setmethod(UnitRange, function (u){return Array(u.length).keys()})

for (let x of iterate(array)){console.log(x)}

let sum = poly.new()
sum.setmethod( _Array, function (a){
    let res = 0
    for (let x of iterate(a)){ res = res + x}
    return res
    }
)
sum.setmethod("number", function (x){ return x})


console.log("sum array: ", sum(array))
console.log("sum number: ", sum(12))

File: lua/src/dispatch.lua
local M = {}

local dynamic_dispatcher = {}

function dynamic_dispatcher.__call(tbl, arg1,  ...)
    local objType = type(arg1)
    if objType == "table" then
        local meta = getmetatable(arg1)
        while meta ~= nil do
            local method = tbl[meta]
            if method ~= nil then
                return method(arg1, ...)
            end
            meta = getmetatable(meta)
        end
    end
    
    local method = tbl[objType]
    if method ~= nil then
        return method(arg1, ...)
    end
    
    local fallback = tbl["any"]
    if fallback ~= nil then
        return fallback(arg1, ...)
    else
        error("Failed to find method")
    end
end

function M.new(init)
    init = init or {}
    setmetatable(init, dynamic_dispatcher)
    return init
end

function M.struct(supertype)
    local retval = {}
    retval.new = function (tbl)
        tbl = tbl or {}
        setmetatable(tbl, retval)
        return tbl
    end
    if supertype ~= nil then setmetatable(retval, supertype) end
    return retval
end


-----------------------------------------------------------------------------

local Array = M.struct()
local Dict = M.struct()

local iterate = M.new {
    [Array] = function (a)
        return ipairs(a)
    end,
    [Dict] = function (d)
        return pairs(d)
    end,
    ["any"] = function (...) return "hello fallback :-)" end,
}


local array = Array.new {2,5,7,9, foo = 12, asdf = 33}

print("iterate as array")
for _, i in iterate(array) do
    print(i)
end

print("")
print("iterate as dict")

local dict = Dict.new {2,5,7,9, foo = 12, asdf = 33}

for k, v in iterate(dict) do
    print(k, "-", v)
end

local sum = M.new {
    [Array] = function (a)
        local retval = 0
        for _, x in ipairs(a) do
            retval = retval + x
        end
        return retval
    end,
    number = function (x) return x end,
}

local UnitRange = M.struct(Array)

local function sumGauss(n) return n * (n + 1) / 2 end


sum[UnitRange] = function (unitRange)
    print("gaussRange")
    return sumGauss(unitRange.stop) - sumGauss(unitRange.start)
end

print("")
print("")
print("testing Sum")
print("")
print("")
print(sum(3))
print("")
print(sum(array))
print("")
print("unitrange:")

local unitrange = UnitRange.new {start = 1, stop = 5}
print(sum(unitrange))
print(iterate({5,6,7}))

local a = {}
local b = {}
b[a] = function (...) return "hi!" end

b.[a]()


